Sources/SwiftParser/Availability.swift

```diff
@@ -262,48 +262,96 @@ extension Parser {
+          } else {
+            trailingComponents.append(versionComponent)
+          }
 
-      if !trailingComponents.isEmpty {
-        unexpectedTrailingComponents = RawUnexpectedNodesSyntax(elements: trailingComponents.compactMap { $0.as(RawSyntax.self) }, arena: self.arena)
-      }
+          if versionComponent.hasError {
+            let unexpectedComponents = components + trailingComponents
+            var unexpectedTokens = [RawSyntax(major)] + unexpectedComponents.map(RawSyntax.init)
+            if let (_, handle) = self.canRecoverTo(anyIn: VersionTupleSyntax.EndOfVersionTupleOptions.self),
+              handle.unexpectedTokens > 0
+            {
+              for _ in 0..<handle.unexpectedTokens {
+                unexpectedTokens.append(RawSyntax(self.consumeAnyToken()))
+              }
+            }
+            return RawVersionTupleSyntax(
+              RawUnexpectedNodesSyntax(unexpectedTokens, arena: self.arena),
+              major: self.missingToken(.integerLiteral, text: nil),
+              components: nil,
+              arena: self.arena
+            )
+          }
+        }
 
-      return RawVersionTupleSyntax(
-        major: major,
-        components: RawVersionComponentListSyntax(elements: components, arena: self.arena),
-        unexpectedTrailingComponents,
-        arena: self.arena
-      )
+        var unexpectedTrailingComponents: RawUnexpectedNodesSyntax?
+
+        if !trailingComponents.isEmpty {
+          unexpectedTrailingComponents = RawUnexpectedNodesSyntax(elements: trailingComponents.compactMap { $0.as(RawSyntax.self) }, arena: self.arena)
+        }
 
+        return RawVersionTupleSyntax(
+          major: major,
+          components: RawVersionComponentListSyntax(elements: components, arena: self.arena),
+          unexpectedTrailingComponents,
+          arena: self.arena
+        )
+      } else {
+        let unexpectedToken = self.eat(.floatingLiteral)
+        return RawVersionTupleSyntax(
+          RawUnexpectedNodesSyntax([unexpectedToken], arena: self.arena),
+          major: self.missingToken(.integerLiteral, text: nil),
+          components: nil,
+          arena: self.arena
+        )
+      }
     } else {
       let major = self.expectDecimalIntegerWithoutRecovery()
-      return RawVersionTupleSyntax(major: major, components: nil, arena: self.arena)
+      if major.isMissing {
+        let unexpectedNodes: RawUnexpectedNodesSyntax?
+        if let (_, handle) = self.canRecoverTo(anyIn: VersionTupleSyntax.EndOfVersionTupleOptions.self),
+          handle.unexpectedTokens > 0
+        {
+          var unexpectedTokens = [RawSyntax]()
+          for _ in 0..<handle.unexpectedTokens {
+            unexpectedTokens.append(RawSyntax(self.consumeAnyToken()))
+          }
+          unexpectedNodes = RawUnexpectedNodesSyntax(elements: unexpectedTokens, arena: self.arena)
+        } else {
+          unexpectedNodes = nil
+        }
```

Comment:
I would prefer this to be a whitelist of characters that we want to consume instead of consuming until the end of a few characters that we currently assume to terminate a `VersionTupleSyntax` (think about how `VersionTupleSyntax` could be used in the future in some context that doesn’t terminate it using `)` or `,`. I.e. I would prefer to just continue eating all period, integer and float literals here.

Also, I think it would be good to share this code snippet between the two `if` branches.

Reply:
I understand. As you mentioned, checking for right paren and commas seems too limited of an approach. 
Although I'm not sure if I fully understood your intention, I have made changes to eat integer and floating literals while checking for periods. 
Thank you for your detailed feedback 😃